### ISSUE TITLE
VST3 output pins don't connect in Ardour - VST2 works

### ISSUE THREAD

**Issue by rexmartin158** [iPlug2/iPlug2#880]
This may only be an issue in one DAW (Ardour) but could be a latent problem in IPlugVST3 that I thought I would bring to light.

Audio output pins do not connect in Ardour.  Sound comes into the plugin but does not return back to the DAW.

With a simple “2-2” only config. loaded into a stereo audio track in Ardour,

```
IsChannelConnected(ERoute::kInput, 0);  both report TRUE
IsChannelConnected(ERoute::kInput, 1);

IsChannelConnected(ERoute::kOutput, 0);  both report FALSE
IsChannelConnected(ERoute::kOutput, 1);

NOutChansConnected() reports "0"
```

This confirms what I am hearing and seeing. Audio is getting into the plugin but is not being returned back to the DAW.

The VST2 version reports TRUE for all 4 pins and works properly.

Other brands of VST3 plugins I have tested work in Ardour so this is not a blanket Ardour VST3 issue. Meanwhile, this same VST3 plugin works in every other DAW I have tested in. So it appears there is something peculiar to Ardour + IPlug.

This problem can be observed with the simple IPlugEffect.vst3 plugin.

What could be the problem here and where to look for the cause/fix?

**Comment by olilarkin** [comment]
Thanks for the report. Probably some sequence of activating I/o buses or attaching their buffers is different in ardour. I did briefly try to look at it but didn't get very far and I noticed that the iPlug2 plugins UI doesn't work well in ardour at all (very slow frame rate. Resizing issues).

**Comment by TBProAudio** [comment]
GUI works fine here (only small resizing issues). 
For VST3 channel contectivity we had inject a fix (same for MixBus):

```
bool VST3::SetBusArrangements:
...
		// Test and find correct IO Config
		if (LegalIOExt(pInputBusArrangements, numInBuses, pOutputBusArrangements, numOutBuses) < 0) { return false; }

		if ((pPlug->GetHost() == kMixbus32C) || (pPlug->GetHost() == kHostArdour))	// TBPROAUDIO 24042022 Mixbus and Ardour seem to choose always the config from init (max bus), so not need to change anything afterwards, without plug does not work
			return true;
...
```

**Comment by rexmartin158** [comment]
Thanks Oli and @TBProAudio for looking into this and TB Pro for the suggested fix.  

I tried to implement the patch that TB Pro suggested above but cannot find any such function "bool VST3::SetBusArrangements".  What file is this located in/where did you add this check?

**Comment by rexmartin158** [comment]
@TBProAudio - thank you for the clarification and additional info.

> we rewrote the whole function to handle various VST3 connection cases

This looks a lot different than the version I currently have.  What prompted so many changes and what else did it fix?  I'm always afraid to mess with something that has been working (Ardour not withstanding).  Have you found this new version solid/no regression issues?

**Comment by olilarkin** [comment]
I have "addressed" this in 7dc5b1f7101a7ef02b7d4afe5e048a856bbd32a7, on master, thanks for the report
